pandas/tests/test_categorical.py

```diff
@@ -1607,6 +1607,113 @@ def test_map(self):
+                                             ordered=False),
+                        'C2': pd.Categorical(list("abaab"),
+                                             categories=list("bac"),
+                                             ordered=True)})
+        # single grouper
+        exp_full = DataFrame({'A': [2.0, 1.0, np.nan],
+                              'B': [25.0, 20.0, np.nan],
+                              'C1': pd.Categorical(list("bac"),
+                                                   categories=list("bac"),
+                                                   ordered=False),
+                              'C2': pd.Categorical(list("bac"),
+                                                   categories=list("bac"),
+                                                   ordered=True)})
+        for col in ['C1', 'C2']:
+            result1 = df.groupby(by=col, as_index=False).mean()
+            result2 = df.groupby(by=col, as_index=True).mean().reset_index()
+            expected = exp_full.reindex(columns=result1.columns)
+            tm.assert_frame_equal(result1, expected)
+            tm.assert_frame_equal(result2, expected)
+
+        # multiple grouper
+        exp_full = DataFrame({'A': [1, 1, 1, 2, 2, 2],
+                              'B': [np.nan, 20.0, np.nan, 25.0, np.nan,
+                                    np.nan],
+                              'C1': pd.Categorical(list("bacbac"),
+                                                   categories=list("bac"),
+                                                   ordered=False),
+                              'C2': pd.Categorical(list("bacbac"),
+                                                   categories=list("bac"),
+                                                   ordered=True)})
+        for cols in [['A', 'C1'], ['A', 'C2']]:
+            result1 = df.groupby(by=cols, as_index=False).mean()
+            result2 = df.groupby(by=cols, as_index=True).mean().reset_index()
+            expected = exp_full.reindex(columns=result1.columns)
+            tm.assert_frame_equal(result1, expected)
+            tm.assert_frame_equal(result2, expected)
+
+    def test_set_index_preserve_dtype(self):
+        # GH13743, GH13854
+        df = DataFrame({'A': [1, 2, 1, 1, 2],
+                        'B': [10, 16, 22, 28, 34],
+                        'C1': pd.Categorical(list("abaab"),
+                                             categories=list("bac"),
+                                             ordered=False),
+                        'C2': pd.Categorical(list("abaab"),
+                                             categories=list("bac"),
+                                             ordered=True)})
+        for cols in ['C1', 'C2', ['A', 'C1'], ['A', 'C2'], ['C1', 'C2']]:
+            result = df.set_index(cols).reset_index()
+            result = result.reindex(columns=df.columns)
+            tm.assert_frame_equal(result, df)
+
+    def test_stack_preserve_dtype(self):
```

Comment:
@jreback 
Just to clarify (test files are a bit confusing):

- `test_stack_preserve_dtype` goes to `tests/frame/test_reshape.py` (other `stack` tests there)
- `test_get_dummies_preserve_dtype` to `tests/test_reshape.py` (class `TestGetDummies` there)

Not sure about:
- `test_set_index_preserve_dtype`: probably to `tests/frame/test_alter_axes.py` (most of `test_set_index*` there) but some tests are also in `test_multilevel.py`
- `test_make_axis_dummies_preserve_dtype`: can also go to `tests/test_reshape.py` (if so then probably in a separate class) but all `make_axis_dummies` tests are in `test_panel.py` (I don't use a Panel here because of the issue with `Panel.to_frame()`)